qiskit_experiments/experiment_data.py

```diff
@@ -12,130 +12,321 @@
+        elif isinstance(data, dict):
             self._add_single_data(data)
         elif isinstance(data, Result):
             self._add_result_data(data)
-        elif isinstance(data, (Job, BaseJob)):
-            self._add_result_data(data.result())
         elif isinstance(data, list):
             for dat in data:
                 self.add_data(dat)
         else:
-            raise QiskitError("Invalid data format.")
+            raise QiskitError(f"Invalid data type {type(data)}.")
+
+    def _add_result_data(self, result: Result) -> None:
+        """Add data from a Result object
 
-    def _add_result_data(self, result: Result):
-        """Add data from qiskit Result object"""
+        Args:
+            result: Result object containing data to be added.
+        """
         num_data = len(result.results)
         for i in range(num_data):
             metadata = result.results[i].header.metadata
-            if metadata.get("experiment_type") == self._experiment._type:
+            if metadata.get("experiment_type") == self._type:
                 data = result.data(i)
                 data["metadata"] = metadata
                 if "counts" in data:
                     # Format to Counts object rather than hex dict
                     data["counts"] = result.get_counts(i)
                 self._add_single_data(data)
 
-    def _add_single_data(self, data):
+    def _add_single_data(self, data: Dict[str, any]) -> None:
         """Add a single data dictionary to the experiment.
+        Args:
+            data: Data to be added.
+        """
+        self._data.append(data)
+
+    def data(self, index: Optional[Union[int, slice, str]] = None) -> Union[Dict, List[Dict]]:
+        """Return the experiment data at the specified index.
 
         Args:
-            data (dict): a data dictionary for a single circuit exection.
+            index: Index of the data to be returned.
+                Several types are accepted for convenience:
+                    * None: Return all experiment data.
+                    * int: Specific index of the data.
+                    * slice: A list slice of data indexes.
+                    * str: ID of the job that produced the data.
+
+        Returns:
+            Experiment data.
+
+        Raises:
+            QiskitError: if index is invalid.
         """
-        # This method is intended to be overriden by subclasses when necessary.
-        if data.get("metadata", {}).get("experiment_type") == self._experiment._type:
-            self._data.append(data)
+        if index is None:
+            return self._data
+        if isinstance(index, (int, slice)):
+            return self._data[index]
+        if isinstance(index, str):
+            return [data for data in self._data if data.get("job_id") == index]
+        raise QiskitError(f"Invalid index type {type(index)}.")
+
+    def add_figure(
+        self,
+        figure: Union[str, bytes],
```

Comment:
It looks like `BaseAnalysis.run` method calls this method with matplotlib `Figure` object rather than bytes. How this code works?

Reply:
It works locally just storing and retrieving the object (though I guess not saving to disk). When connected to a DB API it will have to converted to some sort of file format.